gobblin-data-management/src/main/java/gobblin/data/management/convertion/hive/HiveSource.java

```diff
@@ -0,0 +1,168 @@
+import gobblin.source.workunit.WorkUnit;
+import gobblin.util.HadoopUtils;
+import gobblin.util.io.GsonInterfaceAdapter;
+import gobblin.util.reflection.GobblinConstructorUtils;
+
+
+/**
+ * <p>
+ * A {@link Source} that creates generic workunits for a hive table or a hive partition.
+ * </p>
+ * <ul>
+ *  <li>It uses the {@link HiveDatasetFinder} to find all hive tables and partitions
+ *  <li>The update time of a hive {@link Table} or a hive {@link Partition} if found using {@link HiveUnitUpdateProvider}
+ *  <li>The update time from the previous run is used as previous hive watermark.{@link HiveSourceWatermarker} is
+ *  used to get previous hive watermarks
+ * </ul>
+ *
+ *{@link WorkUnit}s are created if the previous high watermark of a {@link Partition}
+ * or a {@link Table} are lower than the latest update time.
+ *
+ * <p>
+ * The {@link WorkUnit}s contain a serialized json of the {@link HiveTable} or {@link HivePartition} at {@value #HIVE_UNIT_SERIALIZED_KEY}
+ * This is later deserialized by the extractor.
+ * </p>
+ */
+@Slf4j
+@SuppressWarnings("rawtypes")
+public class HiveSource implements Source {
+
+  private static final String OPTIONAL_HIVE_UNIT_UPDATE_PROVIDER_FACTORY_CLASS_KEY =
+      "hive.unit.updateProviderFactory.class";
+  private static final String DEFAULT_HIVE_UNIT_UPDATE_PROVIDER_FACTORY_CLASS = HdfsBasedUpdateProviderFactory.class
+      .getCanonicalName();
+
+  public static final Gson GENERICS_AWARE_GSON = GsonInterfaceAdapter.getGson(Object.class);
+
+  public static final String TABLE_COMPLETE_NAME_KEY = "hive.table.completeName";
+  public static final String HIVE_UNIT_SERIALIZED_KEY = "hive.unit.serialized";
+  public static final String PARTITION_COMPLETE_NAME_KEY = "hive.partition.completeName";
+
+  @Override
+  public List<WorkUnit> getWorkunits(SourceState state) {
+
+    List<WorkUnit> workunits = Lists.newArrayList();
+    try {
+
+      // Initialize
+      HiveSourceWatermarker watermaker = new TableLevelWatermaker(state);
+      HiveUnitUpdateProviderFactory updateProviderFactory =
+          GobblinConstructorUtils.invokeConstructor(HiveUnitUpdateProviderFactory.class, state.getProp(
+              OPTIONAL_HIVE_UNIT_UPDATE_PROVIDER_FACTORY_CLASS_KEY, DEFAULT_HIVE_UNIT_UPDATE_PROVIDER_FACTORY_CLASS));
+      HiveUnitUpdateProvider updateProvider = updateProviderFactory.create(state);
+      IterableDatasetFinder<HiveDataset> datasetFinder = new HiveDatasetFinder(getSourceFs(), state.getProperties());
+
+      // Find hive tables
+      Iterator<HiveDataset> iterator = datasetFinder.getDatasetsIterator();
+
+      while (iterator.hasNext()) {
+        HiveDataset hiveDataset = iterator.next();
+        LongWatermark expectedDatasetHighWatermark = new LongWatermark(new DateTime().getMillis());
+
+        // Create workunits for partitions
+        if (HiveUtils.isPartitioned(hiveDataset.getTable())) {
+          Map<List<String>, Partition> sourcePartitions =
```

Comment:
Yes the keys are not used. I was reusing the existing utility function `HiveUtils.getPartitionsMap`. Added `HiveUtils.getPartitions` that returns a list of partitions